server/channels/store/sqlstore/user_store.go

```diff
@@ -2265,3 +2266,104 @@ func (us SqlUserStore) RefreshPostStatsForUsers() error {
 
 	return nil
 }
+
+func (us SqlUserStore) GetUserReport(
+	sortColumn string,
+	sortDesc bool,
+	pageSize int,
+	lastSortColumnValue string,
+	lastUserId string,
+	startAt int64,
+	endAt int64,
+	roleFilter string,
+	teamFilter string,
+	hasNoTeam bool,
+	hideActive bool,
+	hideInactive bool,
+) ([]*model.UserReportQuery, error) {
+	isPostgres := us.DriverName() == model.DatabaseDriverPostgres
+	selectColumns := []string{"u.Id", "u.LastLogin", "s.LastActivityAt AS LastStatusAt"}
+	for _, column := range model.UserReportSortColumns {
+		selectColumns = append(selectColumns, "u."+column)
+	}
+	if isPostgres {
+		selectColumns = append(selectColumns,
+			"MAX(ps.LastPostDate) AS LastPostDate",
+			"COUNT(ps.Day) AS DaysActive",
+			"SUM(ps.NumPosts) AS TotalPosts",
+		)
+	} else {
+		selectColumns = append(selectColumns,
+			"MAX(p.CreateAt) AS LastPostDate",
+			"COUNT(DATE(FROM_UNIXTIME(p.CreateAt / 1000))) AS DaysActive",
+			"COUNT(p.Id) AS TotalPosts",
+		)
+	}
+
+	sortColumnValue := sortColumn
+	if sortDesc {
+		sortColumnValue += " DESC"
+	}
+
+	query := us.getQueryBuilder().
+		Select(selectColumns...).
+		From("Users u").
+		LeftJoin("Status s ON s.UserId = u.Id").
+		Where(sq.Or{
+			sq.Gt{sortColumn: lastSortColumnValue},
+			sq.And{
+				sq.Eq{sortColumn: lastSortColumnValue},
+				sq.Gt{"u.Id": lastUserId},
+			},
+		}).
+		Where(sq.Expr("u.Id NOT IN (SELECT UserId FROM Bots)")).
+		GroupBy("u.Id", "s.UserId").
+		OrderBy(sortColumnValue, "u.Id")
+
+	if pageSize > 0 {
+		query = query.Limit(uint64(pageSize))
+	}
+
+	if isPostgres {
+		joinQuery := "PostStats ps ON ps.UserId = u.Id"
+		if startAt > 0 {
+			startDate := time.UnixMilli(startAt)
+			joinQuery += fmt.Sprintf(" AND ps.Day >= '%s'", startDate.Format("2006-01-02"))
+		}
+		if endAt > 0 {
+			endDate := time.UnixMilli(endAt)
+			joinQuery += fmt.Sprintf(" AND ps.Day < '%s'", endDate.Format("2006-01-02"))
+		}
```

Comment:
Not a big deal, but any reason we aren't using squirrel for this entire thing?

Reply:
I wanted to avoid duplicating some SQL here. Unless there's something about squirrel where we can add more join parameters afterwards.